docs/apis/subsystems/roles.md

```diff
@@ -0,0 +1,141 @@
+
+Roles can be applied to users in a **context** (eg assign Fred as a teacher in a particular course)
+
+### Context
+
+Here are the possible contexts, listed from the most general to the most specific.
+
+| CONTEXT NAME        | CONTEXT AREA       | CONTEXT LEVEL |
+|---------------------|--------------------|---------------|
+| `CONTEXT_SYSTEM`    | the whole site     | 10            |
+| `CONTEXT_USER`      | another user       | 30            |
+| `CONTEXT_COURSECAT` | a course category  | 40            |
+| `CONTEXT_COURSE`    | a course           | 50            |
+| `CONTEXT_MODULE`    | an activity module | 70            |
+| `CONTEXT_BLOCK`     | a block            | 80            |
+
+An authorized user will be able to assign an arbitrary number of roles to each user in any context.
+
+See [Roles and modules#Context](https://docs.moodle.org/dev/Roles_and_modules#Context) for more information.
+
+### Capabilities
+
+Capabilities can have the following permissions:
+
+1. `CAP_INHERIT`
+2. `CAP_ALLOW`
+3. `CAP_PREVENT`
+4. `CAP_PROHIBIT`
+
+If no permission is defined, then the capability permission is inherited from a context that is more general than the current context. If we define different permission values for the same capability in different contexts, we say that we are overriding the capability in the more specific context.
+
+### Capability conflicts
+
+Since the capabilities in each role could be different, there could be conflict in capabilities.
+
+If we set a `PROHIBIT` on a capability, it means that the capability cannot be overridden and will ALWAYS  have a permission of prevent (deny). Prohibit always wins.   For example, Jeff has a naughty student role that prohibits him from postings in any forums (for the whole site), but he's also assigned a facilitator role in "Science forum" in the course Science and Math 101. Since prohibit always wins, Jeff is unable to post in "Science forum".
+
+## Programming Interface
+
+Although the Roles system may look complicated at first glance, implementing it in Moodle code is fairly simple.
+
+- Examples of functions handling the **Roles** can be found at [`accesslib.php`](https://github.com/moodle/moodle/blob/master/lib/accesslib.php)
+
+- Moodle comes with a list of predefined roles for example Student, Teacher, Course Manager etc. A new role can be created based on
+    those predefined roles using **role archetype:**
+
+```php
+/**
+ * Returns array of all role archetypes.
+ *
+ * @return array
+ */
+function get_role_archetypes() {
+    return array(
+        'manager'        => 'manager',
+        'coursecreator'  => 'coursecreator',
+        'editingteacher' => 'editingteacher',
+        'teacher'        => 'teacher',
+        'student'        => 'student',
+        'guest'          => 'guest',
+        'user'           => 'user',
+        'frontpage'      => 'frontpage'
+    );
+}
+```
```

Comment:
I'd query whether we should put this in in this format. You could instead do something like:

```suggestion

The role archetypes do not change often, and are currently defined as:

- **manager** - A course category role used to manage courses without being directly enrolled in them
- **coursecreator** - ...
- ...

<details>
<summary>Fetching a list of the role archetypes programatically</summary>

In some rare situations you may need to fetch a list of available role archetypes. You can do so using the `get_role_archetypes()` function, for example:

```php title="Fetching a list of role archetypes"
$archetypes = get_role_archetypes();
```

</details>

```